lib/tracker/request-logger.js

```diff
@@ -0,0 +1,281 @@
+   * portion of the message. Defaults to 500.
+   * @param {Number} [options.messageMaxParameterValueLength] The maximum amount of characters of each query parameter
+   * value that will be included in the message.. Defaults to 50.
+   * @param {Number} [options.messageMaxErrorStackTraceLength] The maximum amount of characters of the stack trace
+   * that will be included in the message. Defaults to 200.
+   */
+  constructor(options) {
+    super();
+    if (!options) {
+      throw new errors.ArgumentError('RequestLogger options parameter is required');
+    }
+
+    this._options = options;
+
+    /**
+     * Determines whether it should emit 'normal' events for every EXECUTE, QUERY and BATCH request executed
+     * successfully, useful only for debugging
+     * @type {Boolean}
+     */
+    this.logNormalRequests = this._options.logNormalRequests;
+
+    /**
+     * Determines whether it should emit 'failure' events for every EXECUTE, QUERY and BATCH request execution that
+     * resulted in an error
+     * @type {Boolean}
+     */
+    this.logErroredRequests = this._options.logErroredRequests;
+
+    /**
+     * The object instance that emits <code>'slow'</code>, <code>'large'</code>, <code>'normal'</code> and
+     * <code>'failure'</code> events.
+     * @type {EventEmitter}
+     */
+    this.emitter = new events.EventEmitter();
+  }
+
+  /**
+   * Logs a message when a request execution was deemed slow or too large.
+   * @override
+   */
+  onSuccess(host, query, parameters, queryOptions, requestLength, responseLength, latency) {
+    if (this._options.slowThreshold > 0 && toMillis(latency) > this._options.slowThreshold) {
+      this._logSlow(host, query, parameters, queryOptions, requestLength, responseLength, latency);
+    }
+    else if (this._options.requestSizeThreshold > 0 && requestLength > this._options.requestSizeThreshold) {
+      this._logLargeRequest(host, query, parameters, queryOptions, requestLength, responseLength, latency);
+    }
+    else if (this.logNormalRequests) {
+      this._logNormalRequest(host, query, parameters, queryOptions, requestLength, responseLength, latency);
+    }
+  }
+
+  /**
+   * Logs a message when a request execution was deemed too large.
+   * @override
+   */
+  onError(host, query, parameters, queryOptions, requestLength, err, latency) {
+    if (this._options.requestSizeThreshold > 0 && requestLength > this._options.requestSizeThreshold) {
+      this._logLargeErrorRequest(host, query, parameters, queryOptions, requestLength, err, latency);
+    }
+    else if (this.logErroredRequests) {
+      this._logErrorRequest(host, query, parameters, queryOptions, requestLength, err, latency);
+    }
+  }
+
+  _logSlow(host, query, parameters, queryOptions, requestLength, responseLength, latency) {
+    const message = format('[%s] Slow request, took %d ms (%s): %s', host.address, ~~toMillis(latency),
```

Comment:
Wasn't initially clear to me that what bitwise not operation `~~` was used for, is to round down effectively right?

Reply:
👍 `Math.floor()` is more clear, will change.